include/RAJA/pattern/params/reducer.hpp

```diff
@@ -101,48 +95,99 @@ namespace detail
     unsigned int * device_count = nullptr;
 #endif
 
-    using ARG_TUP_T = camp::tuple<value_type*>;
-    RAJA_HOST_DEVICE ARG_TUP_T get_lambda_arg_tup() { return camp::make_tuple(&val); }
+    // These are types and parameters extracted from this struct, and given to the forall.
+    using ARG_TUP_T = camp::tuple<VType*>;
+    RAJA_HOST_DEVICE ARG_TUP_T get_lambda_arg_tup() { return camp::make_tuple(&valop_m); }
 
     using ARG_LIST_T = typename ARG_TUP_T::TList;
     static constexpr size_t num_lambda_args = camp::tuple_size<ARG_TUP_T>::value ;
   };
 
-} // namespace detail
+  // Partial specialization of Reducer for ValLoc
+  // T must be a ValLoc
+  // TTOp is the template template version of Op, required by ValOp
+  // VType must be ValOp<ValLoc<>,Op>
+  template <typename Op, typename T, template <typename, typename, typename> class TTOp>
+  struct Reducer<Op, T, ValOp<ValLoc<typename T::value_type, typename T::index_type>, TTOp>> : public ForallParamBase {
+    using target_value_type = typename T::value_type; // This should be a basic data type extracted from the T=ValLoc passed in
+    using target_index_type = typename T::index_type;
+    using value_type = T; // Note that this struct operates on ValLoc
+    using op = TTOp<value_type,value_type,value_type>;
+    using VType = ValOp<ValLoc<target_value_type,target_index_type>, TTOp>;
 
-template <template <typename, typename, typename> class Op, typename T>
-auto constexpr Reduce(T *target)
-{
-  return detail::Reducer<Op<T, T, T>, T>(target);
-}
+    Reducer() = default;
 
+    // ValLoc constructor
+    // Note that the passthru variables point to the val and loc within the user defined target ValLoc
+    RAJA_HOST_DEVICE Reducer(value_type *target_in) : valop_m(VType{}), target(target_in), passthruval(&target_in->val), passthruindex(&target_in->loc) {}
 
+    // Pass through constructor for ReduceLoc<>(data, index) case
+    // The passthru variables point to vars defined by the user
+    RAJA_HOST_DEVICE Reducer(target_value_type *data_in, target_index_type *index_in) : valop_m(VType(*data_in, *index_in)), target(&valop_m.val), passthruval(data_in), passthruindex(index_in) {}
 
-namespace detail
-{
+    Reducer(Reducer const &) = default;
+    Reducer(Reducer &&) = default;
+    Reducer& operator=(Reducer const &) = default;
+    Reducer& operator=(Reducer &&) = default;
 
-  //
-  //
-  // Basic ReducerLoc
-  //
-  //
-  template <typename Op, typename T>
-  struct ReducerLoc : public Reducer<Op, T> {
-    using Base = Reducer<Op, T>;
-    using value_type = typename Base::value_type;
-    ReducerLoc(value_type *target_in) {
-      Base::target = target_in;
-      Base::val = value_type(Op::identity());
-    }
+    VType valop_m = VType{};
+
+    value_type *target = nullptr;
```

Comment:
Is this necessary anymore?

Reply:
Yes this is necessary when the user gives us a `ValLoc`, and this target will point to that user-defined `ValLoc`.

Reply:
If we get rid of `value_type *target`, it will need to become a reference within `ValOp valop_m`. Is this what you're suggesting? Doing so is less clutter for us RAJA devs, but I cannot tell if the code is more clear that way.

Reply:
This ties into the getTarget and combineWithTarget method ideas.
We may not need this particular target pointer if the passthrough vars can handle the case of interacting with the user object and the combineWithTarget can handle combining.
Its better to avoid extra vars and more versatile to use a function instead of directly using vars.